awsx/ec2/vpc.ts

```diff
@@ -343,14 +314,119 @@ export class Vpc extends schema.Vpc<VpcData> {
       routes,
       natGateways,
       eips,
-      subnetLayout: subnetLayout,
+      subnetLayout: rewrap(pulumi.output(subnetLayout)),
       privateSubnetIds,
       publicSubnetIds,
       isolatedSubnetIds,
       vpcId,
     };
   }
 
+  // Decide the cidrBlock input parameter for the underlying aws.ec2.Vpc resource.
+  private static decideCidrBlockVpcInput(args: schema.VpcArgs): pulumi.Input<string> | undefined {
+    // Respect the user-provided value, if any.
+    if (args.cidrBlock !== undefined) {
+      return args.cidrBlock;
+    }
+
+    // If the user wants to use an IPAM pool without specifying a cidrBlock, they must also define ipv4netMaskLength
+    // that instructs how the IPAM pool should allocate the cidrBlock. In this case the should not assume any defaults.
+    if (args.ipv4IpamPoolId !== undefined) {
+      return undefined;
+    }
+
+    // Historically this default was used when left unspecified.
+    return "10.0.0.0/16";
+  }
+
+  private decideSubnetSpecs(
+    name: string,
+    cidrBlock: pulumi.Input<string>,
+    subnetStrategy: schema.SubnetAllocationStrategyInputs,
+    availabilityZones: string[],
+    args: {
+      readonly subnetSpecs?: schema.SubnetSpecInputs[];
+      readonly subnetStrategy?: schema.SubnetAllocationStrategyInputs;
+      readonly availabilityZoneCidrMask?: number;
+    },
+  ): {
+    subnetSpecs: SubnetSpecPartial[];
+    subnetLayout: pulumi.Output<schema.ResolvedSubnetSpecOutputs[]>;
+  } {
+    const parsedSpecs = validateAndNormalizeSubnetInputs(
+      args.subnetSpecs,
+      availabilityZones.length,
+    );
+
+    const thisResource = this;
+
+    const subnetSpecs = (() => {
+      // cidrBlock is allocated dynamically but there are no explicit specs; use subnetDistributorNew.
+      if (typeof cidrBlock !== "string" && parsedSpecs === undefined) {
+        return getSubnetSpecsWithPartialCidr(
+          name,
+          cidrBlock,
+          availabilityZones,
+          undefined,
+          args.availabilityZoneCidrMask,
+        );
+      }
+
+      if (parsedSpecs === undefined || subnetStrategy === "Legacy") {
+        if (typeof cidrBlock === "string") {
+          const legacySubnetSpecs = getSubnetSpecsLegacy(
+            name,
+            cidrBlock,
+            availabilityZones,
+            parsedSpecs?.normalizedSpecs,
+          );
+          return legacySubnetSpecs;
+        } else {
+          pulumi.log.warn(
+            `Unable to allocate subnets. Try setting subnetStrategy="Auto"`,
```

Comment:
Should this fail fast?

Reply:
I would vote yes. It will eventually fail and I would rather be told before I deploy than after.

Reply:
Done.